hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/BucketLayoutAwareOMKeyRequestFactory.java

```diff
@@ -0,0 +1,300 @@
+import org.apache.hadoop.ozone.om.request.s3.multipart.S3MultipartUploadCommitPartRequest;
+import org.apache.hadoop.ozone.om.request.s3.multipart.S3MultipartUploadCommitPartRequestWithFSO;
+import org.apache.hadoop.ozone.om.request.s3.multipart.S3MultipartUploadAbortRequest;
+import org.apache.hadoop.ozone.om.request.s3.multipart.S3MultipartUploadAbortRequestWithFSO;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.HashSet;
+
+import static org.apache.hadoop.ozone.om.OzoneManagerUtils.getBucketLayout;
+
+
+/**
+ * Factory class to instantiate bucket layout aware request classes.
+ */
+public final class BucketLayoutAwareOMKeyRequestFactory {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(BucketLayoutAwareOMKeyRequestFactory.class);
+
+  private static final HashMap<String, Class<? extends OMKeyRequest>>
+      OM_KEY_REQUEST_CLASSES = new HashMap<>();
+
+  static {
+    // CreateDirectory
+    addRequestClass(Type.CreateDirectory,
+        OMDirectoryCreateRequest.class,
+        BucketLayout.OBJECT_STORE
+    );
+    addRequestClass(Type.CreateDirectory,
+        OMDirectoryCreateRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+
+    // CreateFile
+    addRequestClass(Type.CreateFile,
+        OMFileCreateRequest.class,
+        BucketLayout.OBJECT_STORE
+    );
+    addRequestClass(Type.CreateFile,
+        OMFileCreateRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED
+    );
+
+    // CreateKey
+    addRequestClass(Type.CreateKey,
+        OMKeyCreateRequest.class,
+        BucketLayout.OBJECT_STORE
+    );
+    addRequestClass(Type.CreateKey,
+        OMKeyCreateRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+
+    // AllocateBlock
+    addRequestClass(Type.AllocateBlock,
+        OMAllocateBlockRequest.class,
+        BucketLayout.OBJECT_STORE
+    );
+    addRequestClass(Type.AllocateBlock,
+        OMAllocateBlockRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED
+    );
+
+    // CommitKey
+    addRequestClass(Type.CommitKey,
+        OMKeyCommitRequest.class,
+        BucketLayout.OBJECT_STORE
+    );
+    addRequestClass(Type.CommitKey,
+        OMKeyCommitRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED
+    );
+
+    // DeleteKey
+    addRequestClass(Type.DeleteKey,
+        OMKeyDeleteRequest.class,
+        BucketLayout.OBJECT_STORE);
+    addRequestClass(Type.DeleteKey,
+        OMKeyDeleteRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+
+    // DeleteKeys
+    addRequestClass(Type.DeleteKeys,
+        OMKeysDeleteRequest.class,
+        BucketLayout.OBJECT_STORE);
```

Comment:
@JyotinderSingh Please add `OMKeysDeleteRequestWithFSO.class` to this set.

Reply:
Done